enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,255 @@
+None
+
+### Implementation Details [optional]
+
+#### _oc login_ command
+
+The oc login command will have a new flag `--auto` which indicates to `oc` that it should use the Oauth2 Authorization
+Flow to login. Currently, when a user enters `oc login $CLUSTER_API_URL`, `oc` attempts to be login on the cluster if
+the cluster has identity provider which
+support [challenge based authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate).
+The new login flow cannot be a fallback if challenge based authentication is unavailable because the user might
+explicitly want to use it to login through code flow with PKCE. For this reason a new flag which indicates to `oc` to
+start the code flow is required.
+
+#### OAuth server
+
+The internal OAuth server uses [OSIN](https://github.com/openshift/osin) as a dependency to provide OAuth2 Authorized
+Grant Flow for a registered client to fetch tokens. It also has support for PKCE(Proof Key for Code Exchange) as
+specified in [rfc7636](https://datatracker.ietf.org/doc/html/rfc7636). _rfc7636_ is an enhancement for public clients,
+i.e. clients which cannot hold client credentials to fetch an access token securely. Similarly `oc` is also a public
+client because it cannot be distributed with client credentials to retrieve the access token and PKCE is designed for
+use cases like this one.
+
+Another change which may be required is to allow the port on the callback URL to be variable when the callback is to
+localhost. This is required because the webserver might not always be able to use the same port to listen on for the
+callback webserver.
+
+#### cluster-authentication-operator
+
+The [cluster-authentication-operator](https://github.com/openshift/cluster-authentication-operator) provisions the
+default [OAuthClient](https://docs.openshift.com/container-platform/4.9/rest_api/oauth_apis/oauthclient-oauth-openshift-io-v1.html)
+resources. It already provisions a client for `oc` but sets the `respondWithChallenges` to _true_ which is unsuitable
+for OAuth Grant Flow. Therefore, another client is required where the `respondWithChallenges` is set to `false` which
+should also be provisioned by the operator.
+
+### Risks and Mitigations
+
+1. There are no changes to existing behavior for users of `oc`. Previous login workflows continue working with the same
+   behavior as in previous versions.
+2. The new login workflow will need to listen on a port for the callback from the OAuth server. This port cannot be
+   fixed because the port may be in-use on the user's machine and the user may not have the permissions to resolve this.
+   Instead, the user should be able to specify an alternate port which is free on the system. In the new client we will
+   allow redirects on the following URLs:
+   * `http://localhost:8280/callback`
+   * `http://localhost:8281/callback`
+   * `http://localhost:8282/callback`
+   * `http://localhost:8283/callback`
+   * `http://localhost:8284/callback`
```

Comment:
As per https://www.rfc-editor.org/rfc/rfc8252.html#section-7.3, you should not use localhost but IPv{4,6} addresses of loopback devices.